File: DP/276_Paint_Fence/java/main.java
class Solution {
  private int[] memo;
 
  public int numWays(int n, int k) {
      // n: number of posts, k: number of colour
      if (n == 0 || k == 0) {
          return 0;
      }
      
      memo = new int[n];
      return totalWays(n, k);
  }
  
  private int totalWays(int posts, int k) {
      if (posts == 1) {
          return k;
      }
      
      if (posts == 2) {
          return k*k;
      }
      
      if (this.memo[posts-1] == 0) {
          this.memo[posts-1] = this.totalWays(posts - 1, k);
      }
      if (this.memo[posts-2] == 0) {
          this.memo[posts-2] = this.totalWays(posts - 2, k);
      }
      
      return (this.memo[posts-1] + this.memo[posts-2]) * (k-1);
  }
}

File: DP/276_Paint_Fence/golang/main.go
var memo []int

func numWays(n int, k int) int {
	memo = make([]int, n)

	if n == 0 || k == 0 {
		return 0
	}

	return totalWays(n, k)
}

func totalWays(posts int, k int) int {
	if posts == 1 {
		return k
	}

	if posts == 2 {
		return k * k
	}

	if memo[posts-1] == 0 {
		memo[posts-1] = totalWays(posts-1, k)
	}

	if memo[posts-2] == 0 {
		memo[posts-2] = totalWays(posts-2, k)
	}

	return memo[posts-1]*(k-1) + memo[posts-2]*(k-1)
}